mm2src/coins/z_coin/storage/walletdb/wasm/storage.rs

```diff
@@ -0,0 +1,1474 @@
+        Ok(Self(
+            IndexedDbBuilder::new(db_id)
+                .with_version(DB_VERSION)
+                .with_table::<WalletDbAccountsTable>()
+                .with_table::<WalletDbBlocksTable>()
+                .with_table::<WalletDbSaplingWitnessesTable>()
+                .with_table::<WalletDbSentNotesTable>()
+                .with_table::<WalletDbTransactionsTable>()
+                .with_table::<WalletDbReceivedNotesTable>()
+                .build()
+                .await?,
+        ))
+    }
+}
+
+#[derive(Clone)]
+pub struct WalletIndexedDb {
+    pub db: SharedDb<WalletDbInner>,
+    pub ticker: String,
+    pub params: ZcoinConsensusParams,
+}
+
+impl<'a> WalletIndexedDb {
+    pub async fn new(
+        ctx: &MmArc,
+        ticker: &str,
+        consensus_params: ZcoinConsensusParams,
+    ) -> MmResult<Self, ZcoinStorageError> {
+        let db = Self {
+            db: ConstructibleDb::new(ctx).into_shared(),
+            ticker: ticker.to_string(),
+            params: consensus_params,
+        };
+
+        Ok(db)
+    }
+
+    #[allow(unused)]
+    async fn lock_db(&self) -> WalletDbRes<WalletDbInnerLocked<'_>> {
+        self.db
+            .get_or_initialize()
+            .await
+            .mm_err(|err| ZcoinStorageError::DbError(err.to_string()))
+    }
+
+    pub async fn is_tx_imported(&self, tx_id: TxId) -> MmResult<bool, ZcoinStorageError> {
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+
+        // Recall where we synced up to previously.
+        let tx_table = db_transaction.table::<WalletDbTransactionsTable>().await?;
+        let index_keys = MultiIndex::new(WalletDbTransactionsTable::TICKER_TXID_INDEX)
+            .with_value(&self.ticker)?
+            .with_value(tx_id.0.to_vec())?;
+        let maybe_tx = tx_table.get_items_by_multi_index(index_keys).await?;
+
+        if !maybe_tx.is_empty() {
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    pub(crate) async fn init_accounts_table(
+        &self,
+        extfvks: &[ExtendedFullViewingKey],
+    ) -> MmResult<(), ZcoinStorageError> {
+        let ticker = self.ticker.clone();
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let walletdb_account_table = db_transaction.table::<WalletDbAccountsTable>().await?;
+
+        // check if account exists
+        let maybe_min_account = walletdb_account_table
+            .cursor_builder()
+            .only("ticker", &ticker)?
+            .bound("height", 0u32, u32::MAX)
+            .open_cursor(WalletDbAccountsTable::TICKER_ACCOUNT_INDEX)
+            .await?
+            .next()
+            .await?;
+        if maybe_min_account.is_some() {
+            return MmError::err(ZcoinStorageError::TableNotEmpty(
+                "Account table is not empty".to_string(),
```

Comment:
you have places, where you repeat 
```rust
ZcoinStorageError::TableNotEmpty( "Account table is not empty".to_string())
```
please use
```rust
#[display(fmt = "Account table is not empty")]
TableNotEmpty,
 ```

Reply:
thanks for the suggestion, the thing this error variant might be use for other tables aside `AccountTable` later

Reply:
Hmm then I suggest to use
```rust
#[display(fmt = "{table_name} table is not empty")]
    TableNotEmpty { 
        table_name: String,
    },
```    
to not repeat "table is not empty" every time

Reply:
great thanks!